### ISSUE TITLE
When changing the book parent on node edit form the available options for the weight should change

### ISSUE THREAD

**Issue by herbdool** [backdrop/backdrop-issues#3636]
**Description of the need**

When changing the book parent on the node edit form the available options for the weight don't change. That is, if the level chosen has more than 30 items, the available weights should be -16 to +16 or more. And if it has fewer than 30 then it's -15 to +15.

![Screenshot from 2019-03-29 14-31-16](https://user-images.githubusercontent.com/1071483/55254942-9ea40280-522f-11e9-9041-928c5aa2e8dc.png)

**Proposed solution**

Either have it update via ajax or change the field type to a number field and let the user choose any weight.

**Comment by herbdool** [comment]
@klonos maybe you're interested in testing.

**Comment by klonos** [comment]
Works as expected @herbdool, and code change is pretty minimal/straight-forward 👍 ...I just left a comment in the PR re max value (which I don't think has any point to be set).

**Comment by herbdool** [comment]
@klonos I'm okay with taking it out, though I did get the max value from core function form_process_weight(). I'm not sure if it matters or not. Want to update the labels?

**Comment by arjunk2004** [comment]
The solution seems to work. However, I don't quite understand why there are no limits on the weight, shouldn't there be a minimum weight?

**Comment by herbdool** [comment]
We don't want fixed minimum and maximum here since there could be larger numbers of items. And it's not easy to make it dynamic, and not really worth the effort and load IMHO. So none is set. This also makes it the same as other edit screens with weights like the taxonomy term edit. @klonos, do you agree?

**Comment by herbdool** [comment]
@klonos your comments suggest you gave it an RTBC if I removed the max value line (which I did) so I've updated the labels.

**Comment by klonos** [comment]
Yep, RTBC 😄 👍

**Comment by quicksketch** [comment]
In testing this I was glad to see that the stand-alone form on `node/%/outline` was also affected by this change, so we don't have a select list in one place but a number field in another.

Merged https://github.com/backdrop/backdrop/pull/2559 into 1.x and 1.12.x. Thanks @herbdool, @klonos, and @arjunk2004!

**Comment by quicksketch** [comment]
Oh, darn. On further investigation we do have a select list instead of a number field on `admin/content/book/%`. Could we get a follow-up to switch that form with the drag and drop elements as well?

**Comment by quicksketch** [comment]
Current outline page even after this PR has been merged:

![image](https://user-images.githubusercontent.com/100206/56857838-e7ca9d80-6926-11e9-95a3-3edf3f38ec07.png)

**Comment by klonos** [comment]
@quicksketch follow-up PR here: https://github.com/backdrop/backdrop/pull/2636

PS: The book module admin pages need some love. They have fallen out of my radar because the module is disabled by default OOTB. I have created a UX meta here: #3736

**Comment by herbdool** [comment]
@klonos there's already a meta here https://github.com/backdrop/backdrop-issues/issues/3639

**Comment by herbdool** [comment]
@quicksketch I don't think we need to change it on the outline page because we decided the select list was still useful for those lists, and that matches the behaviour for menu item lists and taxonomy term lists (versus individual item edits).

**Comment by klonos** [comment]
> @klonos there's already a meta here #3639

**Comment by klonos** [comment]
> I don't think we need to change it on the outline page because we decided the select list was still useful for those lists, and that matches the behaviour for menu item lists and taxonomy term lists (versus individual item edits).

@herbdool I still believe that fields of type weight were an invention/workaround from a time where HTML5 adoption was not good. It's been years now that HTML% number field is used, and it supports setting min/max values (the equivalent of `#delta` used in weight fields).

I have found no evidence/literature online to suggest that drop-downs with a specific set of values offer better UX and/or accessibility over using number fields. That, combined with the known performance issue with using multiple/many drop-downs with many values, tells me that we have a reason to get rid of `weight` fields altogether. I realise that there is a workaround in place that "throttles" this and falls back to number fields, but where you see consistency with menu item lists and taxonomy term lists, I see inconsistency/WTF when a widget that is exposed to users "magically" changes between drop-down (`weight`) and number. Some developers may be familiar with the code, the performance issue behind it, and the workaround in place, but not everyone (and especially non-coders) will understand this.

Another thing that bothers me is this: drops-downs are to be used for specific sets of options, which are known. We cannot leave a `weight` field without setting its `#delta` to a "safe" value (i.e. no way to set it to "unlimited"). Otherwise we are risking possible performance issues when the items. That arbitrary "safe" value we choose seems "magic" though. With the number field, we do not need to worry about anything similar. So fields that use the HTML5 number instead of the Drupal-specific `weight` can be more flexible (leaving min/max unspecified). A somewhat similar use case is #1459 for example.

**Comment by herbdool** [comment]
@klonos maybe we could move your last comment to the other issue as well, since it's about the general arguments for/against  rather than just about book outlines. No need to have the same discussion in two spots.

**Comment by herbdool** [comment]
@klonos I also suggest creating a new issue for your PR as well since it doesn't match the scope of this issue.

**Comment by jenlampton** [comment]
I'm a little lost here on the feedback that's needed, are we changing the weight field on the book outline page from a select element to a number element? 

@klonos it also looks like your latest PR here includes the change that was already committed. Needs a rebase :)
